<a href="https://colab.research.google.com/github/info17tech/Data-Science/blob/main/Multiple_Regression.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import pandas as pd

In [29]:
data = {
    "StudyHours":[2,3,4,5,6,7,8,9,10],
    "Attendance":[60,65,70,75,80,85,90,92,95],
    "Sleep":[6,6,7,7,7,8,8,8,9],
    "Marks":[45,50,58,65,72,80,88,91,96]
}

In [37]:
df = pd.DataFrame(data)
df

,StudyHours,Attendance,Sleep,Marks
0,2,60,6,45
1,3,65,6,50
2,4,70,7,58
3,5,75,7,65
4,6,80,7,72
5,7,85,8,80
6,8,90,8,88
7,9,92,8,91
8,10,95,9,96


## y=b0+b1x1+b2x2+b3x3
### Marks=b0+b1* study hours+b2* attendance+b3* sleephours

Goal is to find b0, b1, b2, b3

In [30]:
from sklearn.linear_model import LinearRegression

# X is independent variables, Y is a dependant variable
X = df[["StudyHours", "Attendance", "Sleep"]]
Y = df["Marks"]

In [31]:
model = LinearRegression()
model.fit(X, Y)

print("Bo-Intercept: ", model.intercept_)
print("Coefficient: ", model.coef_)

Bo-Intercept:  -32.352000000000075
Coefficient:  [1.24       1.148      0.78533333]


### Marks = -32.352000000000075 + 1.24 * study_hours + 1.148 * attendance + 0.78533333 * sleep_hours

β₁ = 1.24 (StudyHours)

If study hours increase by 1 hour, marks increase by 1.24 points, holding attendance and sleep constant.

β₂ = 1.148 (Attendance)

If attendance increases by 1%, marks increase by 1.148 points, holding other variables constant.

β₃ = 0.785 (Sleep)

If sleep increases by 1 hour, marks increase by 0.785 points, holding other variables constant.

Predict the marks

In [32]:
# Method 1
StudyHours = 8
Attendance = 90
SleepHours = 8

Marks = -32.352000000000075 + 1.24 * StudyHours + 1.148 * Attendance + 0.78533333 * SleepHours
print("Predicted Marks: ", Marks)

Predicted Marks:  87.17066663999992


In [33]:
# Method 2
new_student = pd.DataFrame(
    {"StudyHours": [8], "Attendance": [90], "Sleep": [8]}
    )
new_student

,StudyHours,Attendance,Sleep
0,8,90,8


In [34]:
marks_prediction = model.predict(new_student)
print("Predicted Marks: ", marks_prediction)

Predicted Marks:  [87.17066667]


Finding out SSE, SST, SSR for prediction line (R2)

In [35]:
X

,StudyHours,Attendance,Sleep
0,2,60,6
1,3,65,6
2,4,70,7
3,5,75,7
4,6,80,7
5,7,85,8
6,8,90,8
7,9,92,8
8,10,95,9


Calculate the Residuals

In [38]:
df['Predicted_Marks'] = model.predict(X)
df

,StudyHours,Attendance,Sleep,Marks,Predicted_Marks
0,2,60,6,45,43.720000
1,3,65,6,50,50.700000
2,4,70,7,58,58.465333
3,5,75,7,65,65.445333
4,6,80,7,72,72.425333
5,7,85,8,80,80.190667
6,8,90,8,88,87.170667
7,9,92,8,91,90.706667
8,10,95,9,96,96.176000


In [39]:
# Find the residual
df['Residual'] = df['Marks'] - df['Predicted_Marks']
df

,StudyHours,Attendance,Sleep,Marks,Predicted_Marks,Residual
0,2,60,6,45,43.720000,1.280000
1,3,65,6,50,50.700000,-0.700000
2,4,70,7,58,58.465333,-0.465333
3,5,75,7,65,65.445333,-0.445333
4,6,80,7,72,72.425333,-0.425333
5,7,85,8,80,80.190667,-0.190667
6,8,90,8,88,87.170667,0.829333
7,9,92,8,91,90.706667,0.293333
8,10,95,9,96,96.176000,-0.176000


In [40]:
df['Residual_Squared'] = df['Residual'] ** 2
df

,StudyHours,Attendance,Sleep,Marks,Predicted_Marks,Residual,Residual_Squared
0,2,60,6,45,43.720000,1.280000,1.638400
1,3,65,6,50,50.700000,-0.700000,0.490000
2,4,70,7,58,58.465333,-0.465333,0.216535
3,5,75,7,65,65.445333,-0.445333,0.198322
4,6,80,7,72,72.425333,-0.425333,0.180908
5,7,85,8,80,80.190667,-0.190667,0.036354
6,8,90,8,88,87.170667,0.829333,0.687794
7,9,92,8,91,90.706667,0.293333,0.086044
8,10,95,9,96,96.176000,-0.176000,0.030976


Calculate SSE=∑e2

In [48]:
SSE = df['Residual_Squared'].sum()
print('SSE:', SSE)

SSE: 3.565333333333313


Calculate SST

In [49]:
y_mean = df['Marks'].mean()
print('Mean of Y:', y_mean)

Mean of Y: 71.66666666666667


In [50]:
# y - y_mean
df['y_mean_diff'] = df['Marks'] - y_mean
df

,StudyHours,Attendance,Sleep,Marks,Predicted_Marks,Residual,Residual_Squared,y-y_mean
0,2,60,6,45,43.720000,1.280000,1.638400,-26.666667
1,3,65,6,50,50.700000,-0.700000,0.490000,-21.666667
2,4,70,7,58,58.465333,-0.465333,0.216535,-13.666667
3,5,75,7,65,65.445333,-0.445333,0.198322,-6.666667
4,6,80,7,72,72.425333,-0.425333,0.180908,0.333333
5,7,85,8,80,80.190667,-0.190667,0.036354,8.333333
6,8,90,8,88,87.170667,0.829333,0.687794,16.333333
7,9,92,8,91,90.706667,0.293333,0.086044,19.333333
8,10,95,9,96,96.176000,-0.176000,0.030976,24.333333


In [51]:
df['y_mean_diff_sq'] = df['y_mean_diff'] ** 2
df

,StudyHours,Attendance,Sleep,Marks,Predicted_Marks,Residual,Residual_Squared,y-y_mean,y_mean_diff_sq
0,2,60,6,45,43.720000,1.280000,1.638400,-26.666667,711.111111
1,3,65,6,50,50.700000,-0.700000,0.490000,-21.666667,469.444444
2,4,70,7,58,58.465333,-0.465333,0.216535,-13.666667,186.777778
3,5,75,7,65,65.445333,-0.445333,0.198322,-6.666667,44.444444
4,6,80,7,72,72.425333,-0.425333,0.180908,0.333333,0.111111
5,7,85,8,80,80.190667,-0.190667,0.036354,8.333333,69.444444
6,8,90,8,88,87.170667,0.829333,0.687794,16.333333,266.777778
7,9,92,8,91,90.706667,0.293333,0.086044,19.333333,373.777778
8,10,95,9,96,96.176000,-0.176000,0.030976,24.333333,592.111111


In [53]:
SST = df['y_mean_diff_sq'].sum()
print('SST:', SST)

SST: 2713.9999999999995


Calculate SSR

In [55]:
SSR = SST - SSE
print('SSR:', SSR)

SSR: 2710.434666666666


Calculate R square

In [59]:
R_square = SSR / SST
print('R Square:', R_square)

R Square: 0.9986863178580201


In [60]:
# Find R2 of variation percentage
R2_percentage = R_square * 100
print('R2 percentage:', R2_percentage)

R2 percentage: 99.86863178580201


Find R^2 using scikit learn

In [61]:
R_square = model.score(X, Y)
print('R Square:', R_square)

R Square: 0.9986863178580202


Adjusted R^2

adjusted_R2 = 1 - (1-R2)*(n-1)/(n-k-1)

R² from sklearn

n = number of observations

k = number of predictors

In [67]:
n = len(Y)      # number of rows
k = X.shape[1]  # number of predictors (columns)
print('Number of observations:', n)
print('Number of predictors:', k)

adjusted_R2=1-(1-R_square)*(n-1)/(n-k-1)
print('Adjusted R_square:',adjusted_R2)

Number of observations: 9
Number of predictors: 3
Adjusted R_square: 0.9978981085728323
